### DISCUSSION TITLE
Virial computed by deepmd doesn't match lammps

### DISCUSSION THREAD

**Discussion by sniblett402** [deepmodeling/deepmd-kit#230] [Q&A]
Hello,

This might be a feature rather than a bug. I find that the virial computed for a given model by dp_test is very different to the virial computed by lammps for the same structure and same model. Each term in the lammps virial is at least 10x larger than the deepmd version (using metal units in lammps, which I think is correct).

I'm using periodic boundary conditions and I'm aware that lammps uses corrections to the virial formula in that case. Does deepmd use a similar correction, or does it use exactly the formula given in the appendix of the Comp. Phys. Comm.? I'm not sure whether that discrepancy would be large enough to account for the difference I'm seeing though. I'd be grateful for any suggestions of things to investigate!

**Comment by amcadmus** [comment]
Hi 
Could you report in detail how did you calculate virial by lammps? A piece of lammps input script would help.

**Comment by sniblett402** [comment]
Of course, sorry for not including that in the first place. To compute and print the virial tensor I used:

compute virial all pressure NULL virial
thermo_style custom step pe c_virial c_virial[1] c_virial[2] c_virial[3] c_virial[4] c_virial[5] c_virial[6]

And I'm comparing the components of c_virial to the values printed in *.v.out by dp_test.
I'm aware that lammps records its virial components in a different order to dp_test, even accounting for that the values are very different.

Thanks for your attention!

**Comment by amcadmus** [comment]
In lammps the virial keyword in the compute pressure command means the "virial term" in the pressure tensor, i.e. \sum_i r_i f_i / V.  In deepmd-kit, the virial is defined by \sum_i r_i f_i. The difference is 1/V. 

When I consider this difference, the lammps result and dp prediction match.

**Comment by sniblett402** [comment]
Great, thank you for looking into this for me! I was aware of the 1/V factor, but the first time I tried to account for it I must have made a mistake in the unit conversions. I'm getting agreement now. 

Could you please answer me two quick questions, to check I have understood everything?
1) Presumably I should specify virial.raw in eV (i.e. without the volume term)?
2) Does deepmd use a correction for periodic boundary conditions when it computes the virial (like lammps does), or is this correction small enough for short-ranged forces that we can safely neglect it?

**Comment by amcadmus** [comment]
1, Yes, the label in the virial.raw dose not have the volume term.
2, The PBC is treated properly when computing the virial. we convert \sum_i r_i f_i to -\sum_{i\neq j} r_{ij} d E_i / dr_{ij} , by which the PBC can be considered.
